README.md
## Gush

Gush is a rapid git workflow for project maintainers and contributors that works
with backend support for GitHub, Enterprise Github, and more!

Check Gush in action [here](https://vimeo.com/88283752) and [here](https://vimeo.com/85439368)!

[![Build Status](https://travis-ci.org/gushphp/gush.png?branch=master)](https://travis-ci.org/gushphp/gush)
[![Scrutinizer Quality Score](https://scrutinizer-ci.com/g/gushphp/gush/badges/quality-score.png?s=127d28d94969ef366d3bc78808cc89b8eeba51e2)](https://scrutinizer-ci.com/g/gushphp/gush/)
[![Code Coverage](https://scrutinizer-ci.com/g/gushphp/gush/badges/coverage.png?s=674a025b490123ccc3678eb1088d7d0a696004a1)](https://scrutinizer-ci.com/g/gushphp/gush/)
[![SensioLabsInsight](https://insight.sensiolabs.com/projects/160ad92b-b065-482e-9ebd-4cff2b931451/mini.png)](https://insight.sensiolabs.com/projects/160ad92b-b065-482e-9ebd-4cff2b931451)
[![Latest Stable Version](https://poser.pugx.org/gushphp/gush/v/stable.png)](https://packagist.org/packages/gushphp/gush) [![Total Downloads](https://poser.pugx.org/gushphp/gush/downloads.png)](https://packagist.org/packages/gushphp/gush) [![Latest Unstable Version](https://poser.pugx.org/gushphp/gush/v/unstable.png)](https://packagist.org/packages/gushphp/gush) [![License](https://poser.pugx.org/gushphp/gush/license.png)](https://packagist.org/packages/gushphp/gush)
[![Stories in Ready](https://badge.waffle.io/gushphp/gush.png?label=ready)](https://waffle.io/gushphp/gush)
[![Dependency Status](https://www.versioneye.com/php/gushphp:gush/badge.svg)](https://www.versioneye.com/php/gushphp:gush/)

<a href="http://gushphp.org"><img src="https://f.cloud.github.com/assets/328359/1930603/3bd6fec6-7eb0-11e3-9945-f41820336d8c.png" alt="Gush logo"  width="200px"/></a>

Logo courtesy from [@maxakawizard](https://twitter.com/MAXakaWIZARD) and [@kotosharic](https://twitter.com/kotosharic)

*Logo explanation is best depicted from this passage from Psalms 78*:

> True, he struck the rock, and water gushed out, streams flowed abundantly, but can he also give us bread?
> Can he supply meat for his people?” When the Lord heard them, he was furious; his fire broke out against
> Jacob, and his wrath rose against Israel, for they did not believe in God or trust in his deliverance.

There first thread of blood and the following are water gushing out of a rock, connecting the Old Testament
prophecy fulfillment in the New Testament at the cross when Jesus was opened on his side and gushed out
water and blood.

### What is this?

Gush is a php console utility whose intention is to automate common maintainer and contributor tasks.

- creates a Pull Request with a formatted table description of the changes
- creates github release notes
- changes the base branch of a Pull Request
- automates retrieval of issue's message, title and comments as a text
- merges a PR with just the number and includes all github discussion on the commit message
- and much more in the form of intuitive commands!

### Install

Install Gush in two ways:

#### 1) Installing system-wide using composer (recommended)

```bash
$ composer global require gushphp/gush=dev-master
```

If it is the first time you globally install a dependency then make sure
you include `~/.composer/vendor/bin` in $PATH as shown [here](http://getcomposer.org/doc/03-cli.md#global).

### Always keep your Gush installation up to date:

```bash
$ composer global update gushphp/gush
```

#### 2) Installing as a phar file:

```
$ curl -sS http://gushphp.org/installer | php
$ mv gush.phar /usr/local/bin/gush // optionally
```

or

```
$ curl -sS http://gushphp.org/installer | php -- --install-dir=bin
```

### Usage

You may want to start by configuring it:

```bash
$ gush core:configure
Insert your github credentials:
username: cordoval
// ...
Configuration saved successfully.
```

Let's go into a repo, list issues, take one, send a pull request and merge it:

List it:
```bash
$ cd project_directory
$ gush issue:list
 #   State  PR?  Title                                     User       Assignee   Milestone        Labels       Created
 14  open        Tests and Documentation for Commands      cordoval                                            2014-01-10
```

Take it:
```bash
$ gush issue:take 14
$ git branch
* 14-tests-and-documentation-for-commands
```

Do your changes and commit them:
```bash
$ git commit -am "added instructions to use gush"
```

Send PR:
```
$ gush pull-request:create
Bug fix? [y]
// ...
PR Title: Added a bit of documentation under usage
https://github.com/gushphp/gush/pull/94
```

Merge it:
```bash
$ gush pull-request:merge 94
Pull Request successfully merged
```

### Contributions

Please read our [contributor guide](https://github.com/gushphp/gush/blob/master/CONTRIBUTING.md).
See also the [issues list](https://github.com/gushphp/gush/issues?state=open).

Running the test suite:

```bash
$ ./dev
```

### Mailing list and IRC channel

Join the [Mailing List](https://groups.google.com/forum/#!forum/gushphp)
and also on IRC channel #gushphp for discussions and questions.


CHANGELOG.md
CHANGELOG
=========

This changelog references the relevant changes (bug, minor and major feature
additions) done in the corresponding version.

To get the diff for a specific change, go to https://github.com/gushphp/gush/commit/XXX where XXX is the change hash
To get the diff between two versions, go to https://github.com/gushphp/gush/compare/v1.2.0...v1.2.1

* 1.8.0 (2014-06-13)

 * upgraded to symfony components 2.5.*
 * renamed GitHubHelperSubscriber?
 * added TODO.md document

* 1.3.10 (2014-03-12)

 * feature implemented i:assign command
 * feature added symfony/filesystem

* 1.3.9 (2014-03-05)

 * feature added version to generated phar
 * feature implemented EDITOR feature for gush and all commands
 * feature implemented assign command

* 1.3.8 (2014-02-28)

 * feature set GitHub as default
 * feature added fork command and included check on other commands

* 1.3.7 (2014-02-17)

 * bugfix switch git version to package version
 * minor added adapter classes, modifying application to use Adapter instead of GitHub Client

* 1.3.6 (2014-02-12)

 * minor #229 first version of CHANGELOG

* 1.3.5 (2014-02-12)

 * minor #220 improve colors
 * minor #223 add more tests up to 63% coverage
 * minor #224 include schema.json on phar file
 * minor #226 expand names of commands in documentation readme
 * minor #227 plug version and name (Gush) into the Application class
 * minor #206 implemented generate_docu script
 * minor #228 fix insight problems and going back to platinum


installer.php
#!/usr/bin/env php
<?php

/*
 * This file is part of Gush.
 *
 * (c) Luis Cordova <cordoval@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 *
 * Gush installer, heavily inspired by composer installer script
 * credits to http://getcomposer.org
 */

process($argv);

/**
 * processes the installer
 */
function process($argv)
{
    $check      = in_array('--check', $argv);
    $help       = in_array('--help', $argv);
    $force      = in_array('--force', $argv);
    $quiet      = in_array('--quiet', $argv);
    $installDir = false;

    // --no-ansi wins over --ansi
    if (in_array('--no-ansi', $argv)) {
        define('USE_ANSI', false);
    } elseif (in_array('--ansi', $argv)) {
        define('USE_ANSI', true);
    } else {
        // On Windows, default to no ANSI, except in ANSICON and ConEmu.
        // Everywhere else, default to ANSI if stdout is a terminal.
        define('USE_ANSI',
        (DIRECTORY_SEPARATOR == '\\')
            ? (false !== getenv('ANSICON') || 'ON' === getenv('ConEmuANSI'))
            : (function_exists('posix_isatty') && posix_isatty(1))
        );
    }

    foreach ($argv as $key => $val) {
        if (0 === strpos($val, '--install-dir')) {
            if (13 === strlen($val) && isset($argv[$key+1])) {
                $installDir = trim($argv[$key+1]);
            } else {
                $installDir = trim(substr($val, 14));
            }
        }
    }

    if ($help) {
        displayHelp();
        exit(0);
    }

    $ok = checkPlatform($quiet);

    if (false !== $installDir && !is_dir($installDir)) {
        out("The defined install dir ({$installDir}) does not exist.", 'info');
        $ok = false;
    }

    if ($check) {
        exit($ok ? 0 : 1);
    }

    if ($ok || $force) {
        installGush($installDir, $quiet);
        exit(0);
    }

    exit(1);
}

/**
 * displays the help
 */
function displayHelp()
{
    echo <<<EOF
Gush Installer
--------------
Options
--help               this help
--check              for checking environment only
--force              forces the installation
--ansi               force ANSI color output
--no-ansi            disable ANSI color output
--install-dir="..."  accepts a target installation directory

EOF;
}

/**
 * check the platform for possible issues on running gush
 */
function checkPlatform($quiet)
{
    $errors = array();
    $warnings = array();

    $iniPath = php_ini_loaded_file();
    $displayIniMessage = false;
    if ($iniPath) {
        $iniMessage = PHP_EOL.PHP_EOL.'The php.ini used by your command-line PHP is: ' . $iniPath;
    } else {
        $iniMessage = PHP_EOL.PHP_EOL.'A php.ini file does not exist. You will have to create one.';
    }
    $iniMessage .= PHP_EOL.'If you can not modify the ini file, you can also run `php -d option=value` to modify ini values on the fly. You can use -d multiple times.';

    if (ini_get('detect_unicode')) {
        $errors['unicode'] = 'On';
    }

    if (extension_loaded('suhosin')) {
        $suhosin = ini_get('suhosin.executor.include.whitelist');
        $suhosinBlacklist = ini_get('suhosin.executor.include.blacklist');
        if (false === stripos($suhosin, 'phar') && (!$suhosinBlacklist || false !== stripos($suhosinBlacklist, 'phar'))) {
            $errors['suhosin'] = $suhosin;
        }
    }

    if (!function_exists('json_decode')) {
        $errors['json'] = true;
    }

    if (!extension_loaded('Phar')) {
        $errors['phar'] = true;
    }

    if (!extension_loaded('filter')) {
        $errors['filter'] = true;
    }

    if (!extension_loaded('hash')) {
        $errors['hash'] = true;
    }

    if (!extension_loaded('ctype')) {
        $errors['ctype'] = true;
    }

    if (!ini_get('allow_url_fopen')) {
        $errors['allow_url_fopen'] = true;
    }

    if (extension_loaded('ionCube Loader') && ioncube_loader_iversion() < 40009) {
        $errors['ioncube'] = ioncube_loader_version();
    }

    if (version_compare(PHP_VERSION, '5.4', '<')) {
        $errors['php'] = PHP_VERSION;
    }

    if (version_compare(PHP_VERSION, '5.4', '<')) {
        $warnings['php'] = PHP_VERSION;
    }

    if (!extension_loaded('openssl')) {
        $warnings['openssl'] = true;
    }

    if (ini_get('apc.enable_cli')) {
        $warnings['apc_cli'] = true;
    }

    ob_start();
    phpinfo(INFO_GENERAL);
    $phpinfo = ob_get_clean();
    if (preg_match('{Configure Command(?: *</td><td class="v">| *=> *)(.*?)(?:</td>|$)}m', $phpinfo, $match)) {
        $configure = $match[1];

        if (false !== strpos($configure, '--enable-sigchild')) {
            $warnings['sigchild'] = true;
        }

        if (false !== strpos($configure, '--with-curlwrappers')) {
            $warnings['curlwrappers'] = true;
        }
    }

    if (!empty($errors)) {
        out("Some settings on your machine make Gush unable to work properly.", 'error');

        out('Make sure that you fix the issues listed below and run this script again:', 'error');
        foreach ($errors as $error => $current) {
            switch ($error) {
                case 'json':
                    $text = PHP_EOL."The json extension is missing.".PHP_EOL;
                    $text .= "Install it or recompile php without --disable-json";
                    break;

                case 'phar':
                    $text = PHP_EOL."The phar extension is missing.".PHP_EOL;
                    $text .= "Install it or recompile php without --disable-phar";
                    break;

                case 'filter':
                    $text = PHP_EOL."The filter extension is missing.".PHP_EOL;
                    $text .= "Install it or recompile php without --disable-filter";
                    break;

                case 'hash':
                    $text = PHP_EOL."The hash extension is missing.".PHP_EOL;
                    $text .= "Install it or recompile php without --disable-hash";
                    break;

                case 'ctype':
                    $text = PHP_EOL."The ctype extension is missing.".PHP_EOL;
                    $text .= "Install it or recompile php without --disable-ctype";
                    break;

                case 'unicode':
                    $text = PHP_EOL."The detect_unicode setting must be disabled.".PHP_EOL;
                    $text .= "Add the following to the end of your `php.ini`:".PHP_EOL;
                    $text .= "    detect_unicode = Off";
                    $displayIniMessage = true;
                    break;

                case 'suhosin':
                    $text = PHP_EOL."The suhosin.executor.include.whitelist setting is incorrect.".PHP_EOL;
                    $text .= "Add the following to the end of your `php.ini` or suhosin.ini (Example path [for Debian]: /etc/php5/cli/conf.d/suhosin.ini):".PHP_EOL;
                    $text .= "    suhosin.executor.include.whitelist = phar ".$current;
                    $displayIniMessage = true;
                    break;

                case 'php':
                    $text = PHP_EOL."Your PHP ({$current}) is too old, you must upgrade to PHP 5.4 or higher.";
                    break;

                case 'allow_url_fopen':
                    $text = PHP_EOL."The allow_url_fopen setting is incorrect.".PHP_EOL;
                    $text .= "Add the following to the end of your `php.ini`:".PHP_EOL;
                    $text .= "    allow_url_fopen = On";
                    $displayIniMessage = true;
                    break;

                case 'ioncube':
                    $text = PHP_EOL."Your ionCube Loader extension ($current) is incompatible with Phar files.".PHP_EOL;
                    $text .= "Upgrade to ionCube 4.0.9 or higher or remove this line (path may be different) from your `php.ini` to disable it:".PHP_EOL;
                    $text .= "    zend_extension = /usr/lib/php5/20090626+lfs/ioncube_loader_lin_5.3.so";
                    $displayIniMessage = true;
                    break;
            }
            if ($displayIniMessage) {
                $text .= $iniMessage;
            }
            out($text, 'info');
        }

        out('');

        return false;
    }

    if (!empty($warnings)) {
        out("Some settings on your machine may cause stability issues with Gush.", 'error');

        out('If you encounter issues, try to change the following:', 'error');
        foreach ($warnings as $warning => $current) {
            switch ($warning) {
                case 'apc_cli':
                    $text = PHP_EOL."The apc.enable_cli setting is incorrect.".PHP_EOL;
                    $text .= "Add the following to the end of your `php.ini`:".PHP_EOL;
                    $text .= "    apc.enable_cli = Off";
                    $displayIniMessage = true;
                    break;

                case 'sigchild':
                    $text = PHP_EOL."PHP was compiled with --enable-sigchild which can cause issues on some platforms.".PHP_EOL;
                    $text .= "Recompile it without this flag if possible, see also:".PHP_EOL;
                    $text .= "    https://bugs.php.net/bug.php?id=22999";
                    break;

                case 'curlwrappers':
                    $text = PHP_EOL."PHP was compiled with --with-curlwrappers which will cause issues with HTTP authentication and GitHub.".PHP_EOL;
                    $text .= "Recompile it without this flag if possible";
                    break;

                case 'openssl':
                    $text = PHP_EOL."The openssl extension is missing, which will reduce the security and stability of Gush.".PHP_EOL;
                    $text .= "If possible you should enable it or recompile php with --with-openssl";
                    break;

                case 'php':
                    $text = PHP_EOL."Your PHP ({$current}) is quite old, upgrading to PHP 5.4 or higher is recommended.".PHP_EOL;
                    $text .= "Gush works with 5.4+ for most people, but there might be edge case issues.";
                    break;
            }
            if ($displayIniMessage) {
                $text .= $iniMessage;
            }
            out($text, 'info');
        }

        out('');

        return true;
    }

    if (!$quiet) {
        out("All settings correct for using Gush", 'success');
    }

    return true;
}

/**
 * installs gush to the current working directory
 */
function installGush($installDir, $quiet)
{
    $installPath = (is_dir($installDir) ? rtrim($installDir, '/').'/' : '') . 'gush.phar';
    $installDir = realpath($installDir) ? realpath($installDir) : getcwd();
    $file       = $installDir.DIRECTORY_SEPARATOR.'gush.phar';

    if (is_readable($file)) {
        @unlink($file);
    }

    $retries = 3;
    while ($retries--) {
        if (!$quiet) {
            out("Downloading...", 'info');
        }

        // @todo enable https on gush server
        $source = (extension_loaded('openssl') ? 'http' : 'http').'://gushphp.org/gush.phar';

        $errorHandler = new ErrorHandler();
        set_error_handler(array($errorHandler, 'handleError'));

        $fh = fopen($file, 'w');
        if (!$fh) {
            out('Could not create file '.$file.': '.$errorHandler->message, 'error');
        }
        if (!fwrite($fh, file_get_contents($source, false, getStreamContext($source)))) {
            out('Download failed: '.$errorHandler->message, 'error');
        }
        fclose($fh);

        restore_error_handler();
        if ($errorHandler->message) {
            continue;
        }

        try {
            // test the phar validity
            $phar = new Phar($file);
            // free the variable to unlock the file
            unset($phar);
            break;
        } catch (Exception $e) {
            if (!$e instanceof UnexpectedValueException && !$e instanceof PharException) {
                throw $e;
            }
            unlink($file);
            if ($retries) {
                if (!$quiet) {
                    out('The download is corrupt, retrying...', 'error');
                }
            } else {
                out('The download is corrupt ('.$e->getMessage().'), aborting.', 'error');
                exit(1);
            }
        }
    }

    if ($errorHandler->message) {
        out('The download failed repeatedly, aborting.', 'error');
        exit(1);
    }

    chmod($file, 0755);

    if (!$quiet) {
        out(PHP_EOL."Gush successfully installed to: " . $file, 'success', false);
        out(PHP_EOL."Use it: php $installPath", 'info');
    }
}

/**
 * colorize output
 */
function out($text, $color = null, $newLine = true)
{
    $styles = array(
        'success' => "\033[0;32m%s\033[0m",
        'error' => "\033[31;31m%s\033[0m",
        'info' => "\033[33;33m%s\033[0m"
    );

    $format = '%s';

    if (isset($styles[$color]) && USE_ANSI) {
        $format = $styles[$color];
    }

    if ($newLine) {
        $format .= PHP_EOL;
    }

    printf($format, $text);
}

/**
 * function copied from Composer\Util\StreamContextFactory::getContext
 *
 * Any changes should be applied there as well, or backported here.
 *
 * @param string $url URL the context is to be used for
 * @return resource Default context
 * @throws \RuntimeException if https proxy required and OpenSSL uninstalled
 */
function getStreamContext($url)
{
    $options = array('http' => array());

    // Handle system proxy
    if (!empty($_SERVER['HTTP_PROXY']) || !empty($_SERVER['http_proxy'])) {
        // Some systems seem to rely on a lowercased version instead...
        $proxy = parse_url(!empty($_SERVER['http_proxy']) ? $_SERVER['http_proxy'] : $_SERVER['HTTP_PROXY']);
    }

    if (!empty($proxy)) {
        $proxyURL = isset($proxy['scheme']) ? $proxy['scheme'] . '://' : '';
        $proxyURL .= isset($proxy['host']) ? $proxy['host'] : '';

        if (isset($proxy['port'])) {
            $proxyURL .= ":" . $proxy['port'];
        } elseif ('http://' == substr($proxyURL, 0, 7)) {
            $proxyURL .= ":80";
        } elseif ('https://' == substr($proxyURL, 0, 8)) {
            $proxyURL .= ":443";
        }

        // http(s):// is not supported in proxy
        $proxyURL = str_replace(array('http://', 'https://'), array('tcp://', 'ssl://'), $proxyURL);

        if (0 === strpos($proxyURL, 'ssl:') && !extension_loaded('openssl')) {
            throw new \RuntimeException('You must enable the openssl extension to use a proxy over https');
        }

        $options['http'] = array(
            'proxy'           => $proxyURL,
        );

        // enabled request_fulluri unless it is explicitly disabled
        switch (parse_url($url, PHP_URL_SCHEME)) {
            case 'http': // default request_fulluri to true
                $reqFullUriEnv = getenv('HTTP_PROXY_REQUEST_FULLURI');
                if ($reqFullUriEnv === false || $reqFullUriEnv === '' || (strtolower($reqFullUriEnv) !== 'false' && (bool) $reqFullUriEnv)) {
                    $options['http']['request_fulluri'] = true;
                }
                break;
            case 'https': // default request_fulluri to true
                $reqFullUriEnv = getenv('HTTPS_PROXY_REQUEST_FULLURI');
                if ($reqFullUriEnv === false || $reqFullUriEnv === '' || (strtolower($reqFullUriEnv) !== 'false' && (bool) $reqFullUriEnv)) {
                    $options['http']['request_fulluri'] = true;
                }
                break;
        }

        if (isset($proxy['user'])) {
            $auth = urldecode($proxy['user']);
            if (isset($proxy['pass'])) {
                $auth .= ':' . urldecode($proxy['pass']);
            }
            $auth = base64_encode($auth);

            $options['http']['header'] = "Proxy-Authorization: Basic {$auth}\r\n";
        }
    }

    return stream_context_create($options);
}

class ErrorHandler
{
    public $message = '';

    public function handleError($code, $msg)
    {
        if ($this->message) {
            $this->message .= "\n";
        }
        $this->message .= preg_replace('{^copy\(.*?\): }', '', $msg);
    }
}
